### ISSUE TITLE
Microbenchmarks process.kill() does not work

### ISSUE THREAD

**Issue by DanTup** [flutter/flutter#19208]
(splitting this from investigation in #19096 into a more specific issue...) 

In microbenchmarks we have some code that tries to kill the `flutter run` process after it gets its json results:

https://github.com/flutter/flutter/blob/b931640c1d582fe51525b8c1b32641fb13b25d5c/dev/devicelab/lib/tasks/microbenchmarks.dart#L128

This does not work; nothing happens (for my Mac, and on some of the CIs, at least). My suspicion is that this is because the `flutter` executable is a shell script, so this process is the shell and not the VM. The shell doesn't seem to pass `SIGINT` on to the VM correctly - this could be the same issue I [wrestled with in Dart Code](https://github.com/Dart-Code/Dart-Code/issues/907) for a while (I originally thought it was Windows-only).

Assuming it is that, the fix in Dart Code was to read the `pid` of the VM process via observatory and send ` kill` to that too. It's also possible we can kill the whole set of processes with a negative ID, but I think this may require spawning it "detached" which may introduce other issues (on Windows I think it spawns a visible console window - not important in tests, but why I couldn't do it with Dart Code).

**Comment by DanTup** [comment]
@jcollins-g Devon mentioned you might be handy with shell scripts; is this something you've seen before or might have ideas about? (using the VM's `pid` is probably a most foolproof way, but we might have to do some gymnastics to get it in the tests here).

**Comment by jcollins-g** [comment]
Sadly the exec option looks like it won't work since we actually do need to do things after the execution of Dart if it exits with a particular code.

That seems suboptimal.  I'd try really hard to avoid that in the flutter script, maybe by asking the VM folks if there is a way you can detect whether the snapshot is invalid without running it.

**Comment by DanTup** [comment]
> TL;DR: Try adding an 'exec'
> It replaces the executing shell with what you're execing so there's no PID change.

Interesting; I didn't know about that. Though I think this would leave the same issue on Windows and I'm not sure if there's an equiv there (though I haven't confirmed the same issue exists on Windows yet).

Having a way to validate the snapshot is valid (or even just do what that code does - re-generate it if it's not valid, then run it) doesn't seem like a bad idea to me though (cc @devoncarew - wdyt?)

Flutter does have a `--pid-file` switch that writes its pid to a file; so that could be an easy fix (similar to getting the pid from observatory) and would be isolated to the tests; however it does mean if there is other code (now or future) doing something similar, it would need to do something similar.

I might investigate the getting the pid via observatory/pid-file as they seem like the easiest fixes, and I might try testing the `exec` fix (by just removing that snapshot check temporarily) and see if something similar is possible on Windows as a more global fix.

**Comment by jcollins-g** [comment]
@DanTup sounds like a reasonable approach.  I thought this morning about adding some traps to forward signals from the shell to the binary, so that's another way of working around this.  (The most direct path to me seems like the one you mentioned -- having the VM validate whether the snapshot is valid and rebuild it itself if necessary.  It would be really nice if we could do that).  I know almost nothing about Windows process management so I can't speculate on whether this is a problem there or how to solve it, though I'm interested to learn what you find.

**Comment by devoncarew** [comment]
> Having a way to validate the snapshot is valid (or even just do what that code does - re-generate it if it's not valid, then run it) doesn't seem like a bad idea to me though (cc @devoncarew - wdyt?)

I think the only current way is to try and execute it and check for a specific failure code. We'd need changes to snapshot execution to support anything more here; either something like a `--check-snapshot` flag, or having another arg to not actually run the snapshot if it turns out to be valid.

This would likely be generally useful for tools that execute snapshots, like pub. /cc @natebosch @grouma

**Comment by devoncarew** [comment]
And, /cc @rmacnak-google and @a-siva for VM support for snapshot validity detection.

**Comment by rmacnak-google** [comment]
```
#!/bin/bash -e                                                                                                               
VM=$(which dart)
if [ $VM -nt hello.snapshot ]; then
   dart --snapshot=hello.snapshot hello.dart
fi
exec $VM hello.snapshot $@
```

**Comment by jcollins-g** [comment]
@rmacnak-google That seems unlikely to work in the case where the dart executable is reverted to an older version.

**Comment by natebosch** [comment]
> Whichever tool is downloading the SDK should use the current time instead of the archive time, as Git does.

AFAIK that's not the way `apt` works. `brew` does seem to use install time rather than archive time. Many folks also use manual approaches (such as [dvm](https://github.com/cbracken/dvm/blob/master/scripts/dvm#L293)) which are most likely going to use the default arguments when extracting the archive which I believe is usually to use archive times.

I don't think we have scoping to move _everyone_ to some team owned dart installed so I don't believe that solution is feasible.

**Comment by natebosch** [comment]
Note that checking the timestamp on the `dart` binary also won't work for anyone with multiple SDKs switched by updated `$PATH`.

**Comment by devoncarew** [comment]
I think we should have tools in the cli dart vm to validate that a particular snapshot file is supported by that vm, esp. given that the snapshot format doesn't have compatibility guarantees between vm versions.
